### PR TITLE
Bug/perm check cache

### PR DESCRIPTION
**Pull request by tcoz**
Simple Map cache with basic update/read/wipe functions to short-circuit redundant permission calls.

## Summary by CodeRabbit

- **New Features**
    - Introduced a Permission Cache Service to optimize permission management and reduce redundant requests.
    - Updated PermissionService to include caching functions for efficient permission handling.
- **Tests**
    - Added test cases to ensure the correct functionality of the Permission Cache Service.
- **Bug Fixes**
    - Implemented clearing of permissions cache upon user logout to maintain data integrity.

### CODE REVIEW

File: spiffworkflow-frontend/src/services/PermissionCacheService.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+/**
+ * TODO: This can lead to redundant permissions requests if the same individual permission
+ * is requested by different permission sets (any perm in a set not found will
+ * trigger a backend call for the whole set).
+ * This is erring on the side of caution for now, but a more robust individual
+ * checker might be useful.
+ */
+const findPermissionsInCache = (permissionsToCheck: PermissionsToCheck) => {
+  let foundAllPerms = true;
+  Object.keys(permissionsToCheck).forEach((path: string) => {
```
Comment (jbirddog) [inline spiffworkflow-frontend/src/services/PermissionCacheService.ts:27]: Looks like this could use something like `find` or `some` to short circuit.

Comment (tcoz) [reply spiffworkflow-frontend/src/services/PermissionCacheService.ts:27]: I did originally short circuit (just returned false if anything was found) but we have a linter thing that does this:

Screenshot 2024-03-10 at 8 21 01 AM

(I've never seen a return value short circuit forbidden this way). 

Anyhow, I came up with something that should fit the bill (Array.every)

**Comment by burnettk** [comment]
there's a failing cypress test that we need to take a look at here, but otherwise it looks good.

File: spiffworkflow-frontend/src/services/PermissionCacheService.test.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+  it('should update the permissions cache with the provided permissions', () => {
+    const permissionsToCheck = {
+      '/path1': ['read', 'write'],
+      '/path2': ['read'],
+    };
+
+    updatePermissionsCache(permissionsToCheck);
+    expect(inspectPermissionsCache().get('/path1')).toEqual(['read', 'write']);
+    expect(inspectPermissionsCache().get('/path2')).toEqual(['read']);
+  });
```
[the remark this answers is not in the document]
Comment (tcoz) [reply spiffworkflow-frontend/src/services/PermissionCacheService.test.ts]: These are old tests

File: spiffworkflow-frontend/src/services/PermissionCacheService.test.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+  it('should update the permission cache with the provided permissions', () => {
+    const permissionsToCheck = {
+      '/path1': ['read', 'write'],
+      '/path2': ['read'],
+    };
+
+    updatePermissionsCache(permissionsToCheck);
+    expect(inspectPermissionsCache().get('/path1')).toEqual(['read', 'write']);
+    expect(inspectPermissionsCache().get('/path2')).toEqual(['read']);
+  });
```
[the remark this answers is not in the document]
Comment (tcoz) [reply spiffworkflow-frontend/src/services/PermissionCacheService.test.ts]: Old test

**Review by burnettk — APPROVED** [review]

**Pull request merged**